etl/src/main/java/org/pih/hivmigration/etl/sql/LabResultMigrator.groovy

```diff
@@ -0,0 +1,261 @@
+            FROM hivmigration_lab_results r
+                     JOIN hivmigration_patients p ON r.source_patient_id = p.source_patient_id
+                     JOIN hivmigration_encounters e ON r.source_encounter_id = e.source_encounter_id
+            WHERE r.test_type ='ppd';
+            
+            -- Numeric Value
+            SET @ppd_value = (SELECT concept_id FROM concept WHERE uuid='3cecf388-26fe-102b-80cb-0017a47871b2');
+            
+            INSERT INTO obs(person_id, concept_id, encounter_id, obs_datetime, location_id, obs_group_id, value_numeric, creator, date_created, voided, uuid)
+            SELECT p.person_id, @ppd_value, e.encounter_id, r.obs_datetime, e.location_id, r.obs_id, r.value_numeric,1, now(), 0, uuid()
+            FROM hivmigration_lab_results r
+                     JOIN hivmigration_patients p ON r.source_patient_id = p.source_patient_id
+                     JOIN hivmigration_encounters e ON r.source_encounter_id = e.source_encounter_id
+            WHERE r.value_numeric IS NOT NULL AND r.test_type ='ppd';
+            
+            
+            -- Rapid Test
+            --
+            
+            -- Specimen number
+            INSERT INTO obs(person_id, concept_id, encounter_id, obs_datetime, location_id, obs_group_id, value_text, creator, date_created, voided, uuid)
+            SELECT p.person_id, @specimen_number, e.encounter_id, r.obs_datetime, e.location_id, r.obs_id, r.sample_id,1, now(), 0,  uuid()
+            FROM hivmigration_lab_results r
+                JOIN hivmigration_patients p ON r.source_patient_id = p.source_patient_id
+                JOIN hivmigration_encounters e ON r.source_encounter_id = e.source_encounter_id
+                WHERE r.test_type ='tr';
+            
+            -- Coded answer
+            SET @test_rapid_value = (SELECT concept_id FROM concept WHERE uuid='3cd6c946-26fe-102b-80cb-0017a47871b2');
+            SET @positive = (SELECT concept_id FROM concept WHERE uuid='3cd3a7a2-26fe-102b-80cb-0017a47871b2');
+            SET @negative = (SELECT concept_id FROM concept WHERE uuid='3cd28732-26fe-102b-80cb-0017a47871b2');
+            INSERT INTO obs(person_id, concept_id, encounter_id, obs_datetime, location_id, obs_group_id, value_coded, creator, date_created, voided, uuid)
+            SELECT p.person_id, @test_rapid_value, e.encounter_id, r.obs_datetime, e.location_id, r.obs_id,
+                   CASE WHEN ( r.value_boolean=1) THEN (@positive)
+                        ELSE @negative
+                       END AS value_coded
+                    ,1, now(), 0, uuid()
+            FROM hivmigration_lab_results r
+                JOIN hivmigration_patients p ON r.source_patient_id = p.source_patient_id
+                JOIN hivmigration_encounters e ON r.source_encounter_id = e.source_encounter_id
+            WHERE r.test_type ='tr';
+        ''')
+    }
+
```

Comment:
My biggest overall comment from reviewing the above, is that we desperately need some well tested functions that will allow us to migrate obs consistently and repeatedly and which will handle all of the boilerplate that you are copying from query to query.  This will:
a) allow us to have much less buggy code, as we will be writing methods once and testing them without worrying about whether every detail is correct each time
b) allow us to make changes to how we want to handle obs migrations more easily, but just changing the core functions that everything uses, rather than have to change every obs migration statement
c) allow the code to be infinitely more readable and easier to review the meaningful bits - eg. what concepts are we migrating into, what data are we migrating from, etc.

Can you please create some common functions for creating numeric, coded, boolean (which would convert a value of 1 or 't' into the true concept, etc), date, and text obs - and then use those functions here?

Reply:
And by functions, I mean functions or procedures - I assume procedures is more appropriate.

Reply:
I created a procedure. I'm not sure, but it might be less general than you were imagining—it's really just for `hivmigration_lab_results`. We could make it very general by abstracting out that table name, but I think this abstraction is already near the line of being more complicated than it's worth.

Reply:
Yeah, I realized about 5 minutes after I made this comment that this was difficult to make generic at the bulk level, and I don't think what you introduced is really improving things.  If anything, I think this makes things more complicated.  Let's brainstorm a little together later.